s3-source-connector/src/main/java/io/aiven/kafka/connect/s3/source/utils/SourceRecordIterator.java

```diff
@@ -94,29 +91,30 @@ private void nextS3Object() {
     }
 
     private Iterator<S3SourceRecord> createIteratorForCurrentFile() throws IOException {
-        try (S3Object s3Object = s3Client.getObject(bucketName, currentObjectKey);) {
 
-            final Matcher fileMatcher = FILE_DEFAULT_PATTERN.matcher(currentObjectKey);
-            String topicName;
-            int defaultPartitionId;
+        final Matcher fileMatcher = FILE_DEFAULT_PATTERN.matcher(currentObjectKey);
+        String topicName;
+        int defaultPartitionId;
+
+        if (fileMatcher.find()) {
+            // TODO move this from the SourceRecordIterator so that we can decouple it from S3 and make it API agnostic
+            try (S3Object s3Object = sourceClient.getObject(currentObjectKey);) {
 
-            if (fileMatcher.find()) {
                 topicName = fileMatcher.group(PATTERN_TOPIC_KEY);
                 defaultPartitionId = Integer.parseInt(fileMatcher.group(PATTERN_PARTITION_KEY));
```

Comment:
This was a defect/performance improvement I noticved, previously we were downloading every file and then checking if we actually wanted to process it based off information we already had.

@RyanSkraba it looks like this would also be a good place for us to check for task assignment so that we can check if want to process a file before downloading it.

Reply:
For me this would also be in a separate PR. I agree that the predicates are a good idea but the object assignment and SourceRecordIterator refactoring should be self contained PRs.